### PR TITLE
[Jobs] Allow resource requests for driver (entrypoint command)

### LINKED ISSUE #24527: [Jobs] Allow resource requests for driver
### Description

We should be able to set `num_cpus` and `num_gpus` for the driver script.  See https://github.com/ray-project/ray/issues/24455#issuecomment-1119063830

### Use case

Users may want to do computationally intensive tasks directly within their driver tasks.  Currently we tell users to move all their computation inside tasks and actors, but this is a limitation.

### PR DESCRIPTION
**Pull request by architkulkarni**
- Allows `num_cpus`, `num_gpus` and `resources` to be reserved for the job entrypoint command.  This is achieved by passing them in as `.options()` to the `JobSupervisor` actor.
- Adds a friendly error message if the JobSupervisor cannot be scheduled

TODO before merge:
- [x] Update docs
- [x] Add tests
- [x] Bump API version number if needed and check backwards compatibility test (backwards compatible because we're adding optional fields, so no API version bump needed)
- [x] Make CUDA_VISIBLE_DEVICES workaround opt-in only
    - [x] Add test

## Related issue number
~~Reverts most of https://github.com/ray-project/ray/pull/24546 as it is no longer necessary.~~

Closes #24527 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by architkulkarni** [comment]
@simon-mo @tchordia I still need to add/update tests but this is otherwise ready for review

**Comment by sihanwang41** [comment]
> @simon-mo @tchordia @sihanwang41 I still need to add/update tests but this is otherwise ready for review

LGTM, after you add tests, I can approve it.

**Review by simon-mo — APPROVED** [review]

**Comment by wuisawesome** [comment]
OOC why are we doing this? (none of the other ways of running a workload allow the driver/entry point to reserve ray resources right?)

**Comment by architkulkarni** [comment]
@wuisawesome It's a popular request from users.  You're right that none of the other ways of running a workload allow this though. 

I think users can have computationally intensive work that isn't happening inside a task or actor (maybe not even a Python script) and it's awkward for them to wrap it in an actor.

**Comment by architkulkarni** [comment]
Maybe a more compelling reason is it fixes a regression from `ray submit` when trying to run a script that uses GPUS: https://github.com/ray-project/ray/pull/24546

**Comment by architkulkarni** [comment]
@amogkam does the approach taken here make sense as a fix for https://github.com/ray-project/ray/issues/24455? We're just allowing the user to set `num_gpus` for the supervisor actor, and removing the workaround from Ed's PR https://github.com/ray-project/ray/pull/24546.  

We don't need to do anything else right? Are there any concerns with nested tasks and actors?

**Comment by amogkam** [comment]
@architkulkarni yep this will work, but the main thing is that users need to be aware of this `JobSupervisorActor`. Is that the plan?

**Comment by architkulkarni** [comment]
@amogkam Good point, they don't have to think about specifying `num_gpus` with the legacy `ray submit`, so ideally they shouldn't have to think about it with `ray job submit`.  User's shouldn't be aware that there's a JobSupervisorActor.

In that case, perhaps it's better to just leave Ed's workaround in this PR?

**Comment by amogkam** [comment]
Yeah that makes sense to me

File: dashboard/modules/job/sdk.py
Comment (architkulkarni) [inline dashboard/modules/job/sdk.py:209]: @simon-mo Can you review this and say if it's a reasonable approach to ensure forward compatibility? It feels a bit too "manual" for what seems like a very generic problem, so I wonder if there's a better way I'm missing.

WIthout this, old servers populate the json into a `JobSubmitRequest` object to validate it, so it fails when it encounters the unrecognized key `num_cpus`.

We could do this key-removing logic on the server side instead, but then we'd only have forward compatibility starting with servers running Ray 2.2+.  With the current approach, it works with older servers right off the bat.

**Comment by architkulkarni** [comment]
>  Good point, they don't have to think about specifying `num_gpus` with the legacy `ray submit`, so ideally they shouldn't have to think about it with `ray job submit`. User's shouldn't be aware that there's a JobSupervisorActor.
> 
> In that case, perhaps it's better to just leave Ed's workaround in this PR?

cc @edoakes, interested in your thoughts on this

File: dashboard/modules/job/sdk.py
Comment (architkulkarni) [reply dashboard/modules/job/sdk.py:209]: cc @edoakes also

**Comment by edoakes** [comment]
@architkulkarni my thinking is that by default users should continue to get the behavior they do today, but if they opt-in to specifying resources for the driver they should get the same behavior as they do for any regular actor. This would also avoid breaking any compatibility.

To be specific:

- Leave the existing workaround in if none of `num_cpus`, `num_gpus`, or `resources` are specified.
- If any of the above are specified, simply pass them to the actor and call it a day.

With this behavior users don't need to know the specifics of the `JobSupervisorActor`, just that they get the same semantics as specifiying resources for an actor if they opt-in by providing a resource constraint.

File: dashboard/modules/job/common.py
Comment (edoakes) [inline dashboard/modules/job/common.py:103]: it'd be very nice to add the required resources and currently available resources in the cluster to this message (similar to the one that we log in the serve controller)

File: dashboard/modules/job/sdk.py
Comment (edoakes) [inline dashboard/modules/job/sdk.py:119]: is this relevant to the resource request change or just a fly-by fix?

File: dashboard/modules/job/sdk.py
Comment (edoakes) [reply dashboard/modules/job/sdk.py:211]: I wish there was an option to pass to `dataclasses` for this but that doesn't appear to be the case. This approach seems ok to me. Could you please add a unit test for the behavior to ensure it isn't broken in the future?

I also see no reason not to do it both on the client and server to be extra defensive. Maybe you can kill two birds with one stone and add a utility that's used both on client & server, then add a unit test for it.

**Review by edoakes — COMMENTED** [review]
Implementation looks good. My only major concern is about the API. I think it would be easy for users to mistake the options for applying to the entire job instead of only the driver (which has also been a feature request in the past). Should we modify the option names to account for this? e.g., `driver_num_cpus`

File: dashboard/modules/job/sdk.py
Comment (architkulkarni) [reply dashboard/modules/job/sdk.py:119]: Ah, the latter, it's just for the `submission_id` change which happened a while ago.  Ran into this during the UX assessment

File: dashboard/modules/job/sdk.py
Comment (architkulkarni) [reply dashboard/modules/job/sdk.py:211]: Makes sense, I'll add it to the server side as well and add a unit test

**Comment by architkulkarni** [comment]
> Implementation looks good. My only major concern is about the API. I think it would be easy for users to mistake the options for applying to the entire job instead of only the driver (which has also been a feature request in the past). Should we modify the option names to account for this? e.g., `driver_num_cpus`

Makes sense! @edoakes would you okay with `entrypoint_num_cpus`?   The reason is we already have `entrypoint` as an argument, and a job can have 0 to N drivers (though the common case is 1 driver). The downside is it's a longer name.

**Comment by edoakes** [comment]
Good idea, I also like that better than my suggestion because it is more consistent with the current API and requires the user to understand one fewer concept.

File: dashboard/modules/job/common.py
Comment (architkulkarni) [reply dashboard/modules/job/common.py:103]: Agreed. I didn't see an quick way to add this, but made a followup issue here https://github.com/ray-project/ray/issues/29921

**Review by edoakes — COMMENTED** [review]
@architkulkarni could you please also update the documentation to include a short section about this? It should also specify the default behavior. I believe we already have a short blurb about that, let's coalesce them.

**Comment by architkulkarni** [comment]
@edoakes Good point, I couldn't find an existing blurb so I just added a section to the doc, please take a look.

**Comment by architkulkarni** [comment]
@stephanie-wang can you please review the new doc as codeowner?

File: doc/source/cluster/running-applications/job-submission/sdk.rst
Comment (sihanwang41) [inline doc/source/cluster/running-applications/job-submission/sdk.rst:208]: dump question: it means we will use as much resource as the machine has? (answer can be added to the doc directly :))

Comment (architkulkarni) [reply doc/source/cluster/running-applications/job-submission/sdk.rst:208]: Good question, that's what it means. It's the same as specifying in `@ray.remote()` for a task/actor.  These are logical CPUs /GPUs, so they're just counters used so Ray knows where to schedule and not to schedule too many tasks/actors on the same node, it doesn't actually limit the CPU/GPU utilization.

I'll add it to the doc

**Review by edoakes — APPROVED** [review]

**Review by DmitriGekhtman — APPROVED** [review]

**Comment by DmitriGekhtman** [comment]
PR's a little on the long side.

**Comment by architkulkarni** [comment]
Lint failure `test_event` was broken on master and unrelated
RLlib memory_leak_a3c flaky on master and unrelated
Windows: tests:test_cancel flaky on master and unrelated

**Pull request merged**
